include/fmt/core.h

```diff
@@ -645,12 +662,19 @@ struct is_contiguous<std::basic_string<Char>> : std::true_type {};
 
 namespace detail {
 
+#if FMT_CTF_CUSTOM_BACK_INSERT_ITERATOR
+template <typename Container> class back_insert_iterator;
+#else
+template <typename Container>
+using back_insert_iterator = std::back_insert_iterator<Container>;
+#endif
```

Comment:
Add a new iterator if needed but please leave `std::back_insert_iterator` handling as before.

Reply:
I'm not sure what do you mean by `std::back_insert_iterator` handling, because this iterator is in the core of {fmt}. For me, this iterator is similar to `string_view`. That's why I propose to use a custom one all over the library, but only when an `std::` one doesn't meet some requirements. Could you give me more details, please?

Reply:
Consider the following example that uses format string compilation (https://godbolt.org/z/E9WYss):

```c++
#include <fmt/compile.h>

int main() {
  char buffer[10];
  fmt::format_to(buffer, FMT_COMPILE("{}"), 42);
  return buffer[0];
}
```

This does almost everything at compile time already only calling the `write` function at runtime:

```asm
main:
        sub     rsp, 24
        mov     esi, 42
        lea     rdi, [rsp+6]
        call    char* fmt::v7::detail::write<char, char*, int, 0>(char*, int)
        movsx   eax, BYTE PTR [rsp+6]
        add     rsp, 24
        ret
```

So you shouldn't change any buffers which are only relevant to the type-erased API. Instead, you should make `write` constexpr (which I guess you already do) and integrate with the format string compilation API (`FMT_COMPILE`). `back_insert_iterator` also becomes irrelevant since you can provide any iterator to the `write` functions.

Reply:
Ok, got it, I definitely missed that. I will try to rework the current implementation using compile-time API then.

Reply:
I'll close the issue for now then.

Reply:
Looks like I have to use `FMT_COMPILE` explicitly in calls for `format_to` with using compile-time API under the hood. Otherwise, an `format_to` overload from core.h would be used and I'm unable to use compile-time API there or fwd anything from compile-time API (because it does not work with constexpr). The use of `FMT_COMPILE` in `format_to` usages is not what I want to be honest. Maybe there is some solution for that but I don't have it yet.

Reply:
I mean it forces the end-users to use `FMT_COMPILE` in `format_to` usages when they want to format text at compile-time.

Reply:
I think you can make `format_to` consteval to avoid using the macro. Macro is only there for compatibility with older compilers.

Reply:
@vitaut did you mean that - https://godbolt.org/z/b4v4xj ?

Reply:
I don't think you can overload on constevalness so it will have to be a separate function like `format_to`. A possibly better option is to use `is_constant_evaluated` in `format_to` in which case you don't need another function.

Reply:
Which `format_to` do you mean? Because if we are talking about the one in core.h then, as I said before, I'm unable to use anything from compile-time API there.

Reply:
No, it doesn't make much sense to change the type-erased functions in `fmt/core.h` or `fmt/format.h` for reasons discussed earlier. Only the `format_to` in `fmt/compile.h` needs to be changed or a new variant of it added if that proves too difficult.

Reply:
Yes, looks like:

- another function like `format_to` but with a different name
- using `FMT_COMPILE` in `format_to` to force the one that uses `compiled_string`

are only options here. Both of these options mean that using the library at compile-time will be slightly different from using it at runtime. Both of these options force to not use `format_to` from core.h because otherwise, it's not clear to the compiler  which version to choose.

Anyway, I will continue with `FMT_COMPILE` approach to present an updated implementation of compile-time formatting as soon as possible.

Reply:
@vitaut Should I open a new PR or force-push to this?

Reply:
I suggest opening a new one.

Reply:
Done https://github.com/fmtlib/fmt/pull/2019